java/pattern/constraint/ThingConstraint.java

```diff
@@ -426,7 +432,7 @@ public boolean equals(final Object o) {
             if (this == o) return true;
             if (o == null || getClass() != o.getClass()) return false;
             final Has that = (Has) o;
-            return (this.type.equals(that.type) && this.attribute.equals(that.attribute));
+            return Objects.equals(this.type, that.type) && Objects.equals(this.attribute, that.attribute);
```

Comment:
Why did we change `this.attribute.equals(that.attribute)` to using `Objects.equals()`? `Objects.equals()` is only used when the object you want to compare could possible be `null`. Which is why we should only use it for `this.type`.

Reply:
Because we thought it looked neater to use the same method for this.attribute and this.type. Makes it easier to read. You are indeed correct we can remain with .equals for attribute.

Reply:
Happy to change it back to.equals though